### PR TITLE
Remove extensions v1beta1 from addon manager and kubectl prune

### LINKED ISSUE #98723: Addon manager defaults prune whitelist to extensions/v1beta1 Ingress
**What happened**:

Sweeping uses of extensions/v1beta1 in preparation for [stopping serving this API group in 1.22](https://github.com/kubernetes/kubernetes/issues/43214), the addon manager stood out as using extensions/v1beta1 Ingress by default:

```
		*pruneResources = []pruneResource{
			{"", "v1", "ConfigMap", true},
			{"", "v1", "Endpoints", true},
			{"", "v1", "Namespace", false},
			{"", "v1", "PersistentVolumeClaim", true},
			{"", "v1", "PersistentVolume", false},
			{"", "v1", "Pod", true},
			{"", "v1", "ReplicationController", true},
			{"", "v1", "Secret", true},
			{"", "v1", "Service", true},
			{"batch", "v1", "Job", true},
			{"batch", "v1beta1", "CronJob", true},
			{"extensions", "v1beta1", "Ingress", true},
			{"apps", "v1", "DaemonSet", true},
			{"apps", "v1", "Deployment", true},
			{"apps", "v1", "ReplicaSet", true},
			{"apps", "v1", "StatefulSet", true},
		}
```

```
  KUBECTL_PRUNE_WHITELIST=(
    core/v1/ConfigMap
    core/v1/Endpoints
    core/v1/Namespace
    core/v1/PersistentVolumeClaim
    core/v1/PersistentVolume
    core/v1/Pod
    core/v1/ReplicationController
    core/v1/Secret
    core/v1/Service
    batch/v1/Job
    batch/v1beta1/CronJob
    apps/v1/DaemonSet
    apps/v1/Deployment
    apps/v1/ReplicaSet
    apps/v1/StatefulSet
    extensions/v1beta1/Ingress
  )
```

**What you expected to happen**:

All components shipped in 1.21 must use networking.k8s.io/v1 Ingress in order to work properly against a 1.22 API server.

/sig networking

cc @MrHohn per cluster/addons/addon-manager/OWNERS
cc @robscott 

I'm not sure what the addon manager does if given a prune list of networking.k8s.io/v1 Ingress objects while still using a manifest containing extensions/v1 Ingress objects... this needs to be tested.

/milestone v1.21
/priority important-soon
/lifecycle frozen

### PR DESCRIPTION
**Pull request by rikatz**
**What type of PR is this?**

Cleanup before v1.22 shipping, removing extensions/v1beta1 Ingress from addon manager

/kind cleanup

**What this PR does / why we need it**: Before shipping v1.22 we need to get rid of any extensions/v1beta1 reference in the code and auxiliary tools, and this is needed to cleanup Addon manager

**Which issue(s) this PR fixes**:
Fixes #98723

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
xref: #43214  #72203 

**Still need to make tests about addon manager "what the addon manager does if given a prune list of networking.k8s.io/v1 Ingress objects while still using a manifest containing extensions/v1 Ingress objects" per Jordan comment**

### CODE REVIEW

**Comment by rikatz** [comment]
/unassign @MrHohn @shiywang 

When it's ready will auto-cc again :)

**Comment by rikatz** [comment]
@liggitt about the tests:

* Created the addons with the previous addonmanager + previous kubectl in a v1.20 cluster:

```
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist extensions/v1beta1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-11T18:26:00-03:00 with ADDON_CHECK_INTERVAL_SEC=60 ==
INFO: == Default service account in the kube-system namespace has token default-token-qqzql ==
find: ‘/etc/kubernetes/admission-controls’: No such file or directory
INFO: == Entering periodical apply loop at 2021-02-11T18:26:00-03:00 ==
Error from server (NotFound): error when creating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard/dashboard.yaml": namespaces "kubernetes-dashboard" not found
Error from server (NotFound): error when creating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard/dashboard.yaml": namespaces "kubernetes-dashboard" not found
Error from server (NotFound): error when creating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard/dashboard.yaml": namespaces "kubernetes-dashboard" not found
Error from server (NotFound): error when creating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard/dashboard.yaml": namespaces "kubernetes-dashboard" not found
INFO: == Kubernetes addon ensure completed at 2021-02-11T18:26:01-03:00 ==
INFO: == Reconciling with deprecated label ==
Error from server (NotFound): error when creating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard/dashboard.yaml": namespaces "kubernetes-dashboard" not found
INFO: == Reconciling with addon-manager label ==
namespace/kubernetes-dashboard created
serviceaccount/kubernetes-dashboard created
role.rbac.authorization.k8s.io/kubernetes-dashboard created
clusterrole.rbac.authorization.k8s.io/kubernetes-dashboard created
rolebinding.rbac.authorization.k8s.io/kubernetes-dashboard created
clusterrolebinding.rbac.authorization.k8s.io/kubernetes-dashboard created
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/kubecost-ingress created
```

* Changed kubectl + addonmanager to use networking.k8s.io/v1 Ingress object (I've added an extra line on kube-addons.sh to print my kubectl version and be sure this is the one being used)

```
Client Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-alpha.3.146+f41def8d76660c-dirty", GitCommit:"f41def8d76660c94c4eca483ec6d3edde5deac89", GitTreeState:"dirty", BuildDate:"2021-02-11T21:27:56Z", GoVersion:"go1.15.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-21T01:11:42Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist networking.k8s.io/v1/Ingress  ==
INFO: == Reconciling with deprecated label ==
service/kubernetes-dashboard unchanged
INFO: == Reconciling with addon-manager label ==
namespace/kubernetes-dashboard unchanged
serviceaccount/kubernetes-dashboard unchanged
role.rbac.authorization.k8s.io/kubernetes-dashboard unchanged
clusterrole.rbac.authorization.k8s.io/kubernetes-dashboard unchanged
rolebinding.rbac.authorization.k8s.io/kubernetes-dashboard unchanged
clusterrolebinding.rbac.authorization.k8s.io/kubernetes-dashboard unchanged
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/kubecost-ingress unchanged
```

So I guess there's no impact doing this. Do you want me to test on some older cluster as well?

**Comment by liggitt** [comment]
the scenarios I was envisioning were:

1. old addon manager + extensions/v1beta1 manifests → old addon manager + networking.k8s.io/v1 manifests
2. old addon manager + extensions/v1beta1 manifests → new addon manager + extensions/v1beta1 manifests
3. new addon manager + extensions/v1beta1 manifests → new addon manager + networking.k8s.io/v1 manifests
4. old addon manager + new manifests → new addon manager + networking.k8s.io/v1 manifests

If each of those transitions works properly (manifests are applied correctly, existing objects aren't incorrectly deleted), then no matter what path is taken from old+old → new+new, we should be safe.

**Comment by liggitt** [comment]
(I'm assuming all scenarios run on 1.19+ which has v1 Ingress available)

**Comment by rikatz** [comment]
OK, I'll deal with this tomorrow and let you know about the scenarios results :)

**Comment by rikatz** [comment]
* Scenario 1: old addon manager + extensions/v1beta1 manifests → old addon manager + networking.k8s.io/v1 manifests

If only the group/apiVersion is updated, a validation error occurs because of Ingress object syntax change (2021-02-12T16:32:32-03:00).

Then, after proper updating/correcting the object syntax, it gets re-applied (probably because the syntax change also), and not changed anymore

So for this test:
* If only the manifest is migrated, it should be properly migrated to the correct object syntax
* It WILL get re-applied once

Below are the logs for this scenario, moving to the next one

```
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist extensions/v1beta1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-12T16:27:36-03:00 with ADDON_CHECK_INTERVAL_SEC=60 ==
INFO: == Default service account in the kube-system namespace has token default-token-qqzql ==
INFO: == Entering periodical apply loop at 2021-02-12T16:27:37-03:00 ==
INFO: == Reconciling with addon-manager label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/dashboard created
INFO: == Kubernetes addon reconcile completed at 2021-02-12T16:27:38-03:00 ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Reconciling with addon-manager label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/dashboard unchanged
INFO: == Kubernetes addon reconcile completed at 2021-02-12T16:28:40-03:00 ==

----- Change manifest to networking.k8s.io/v1 ------
INFO: Leader is 
error validating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard.yaml": error validating data: ValidationError(Ingress.spec): unknown field "backend" in io.k8s.api.networking.v1.IngressSpec; if you choose to ignore these errors, turn validation off with --validate=false
INFO: == Kubernetes addon ensure completed at 2021-02-12T16:32:32-03:00 ==
INFO: == Reconciling with deprecated label ==
service/kubernetes-dashboard unchanged
error: error validating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard.yaml": error validating data: ValidationError(Ingress.spec): unknown field "backend" in io.k8s.api.networking.v1.IngressSpec; if you choose to ignore these errors, turn validation off with --validate=false
INFO: == Reconciling with addon-manager label ==
error: error validating "/home/rkatz/archive/rkatz/codes/kubernetes/cluster/addons/addon-manager/bla/dashboard.yaml": error validating data: ValidationError(Ingress.spec): unknown field "backend" in io.k8s.api.networking.v1.IngressSpec; if you choose to ignore these errors, turn validation off with --validate=false

----- Correct the ingress specification -----
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist extensions/v1beta1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-12T16:36:18-03:00 with ADDON_CHECK_INTERVAL_SEC=60 ==
INFO: == Default service account in the kube-system namespace has token default-token-qqzql ==
INFO: == Entering periodical apply loop at 2021-02-12T16:36:18-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T16:36:19-03:00 ==
INFO: == Reconciling with deprecated label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Reconciling with addon-manager label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Kubernetes addon reconcile completed at 2021-02-12T16:36:21-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T16:37:18-03:00 ==
INFO: == Reconciling with deprecated label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Reconciling with addon-manager label ==
ingress.networking.k8s.io/dashboard unchanged
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Kubernetes addon reconcile completed at 2021-02-12T16:37:21-03:00 ==
```

**Comment by liggitt** [comment]
> If only the group/apiVersion is updated, a validation error occurs because of Ingress object syntax change

for all the "new manifest" scenarios, I was assuming a valid new manifest, not just a find/replace of the apiVersion

**Comment by rikatz** [comment]
> > If only the group/apiVersion is updated, a validation error occurs because of Ingress object syntax change
> 
> for all the "new manifest" scenarios, I was assuming a valid new manifest, not just a find/replace of the apiVersion

Great, so yeap, I'll finish the tests here and create a matrix of results. For now will post the logs just for historic context :) Want me to ping you on slack when I finish this, so you can for now disregard GH notifications?

**Comment by liggitt** [comment]
> Want me to ping you on slack when I finish this, so you can for now disregard GH notifications?

👍

**Comment by rikatz** [comment]
* Scenario 2: old addon manager + extensions/v1beta1 manifests → new addon manager + extensions/v1beta1 manifests
Everything works as expected, nothing gets pruned / reapplied:

```
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist extensions/v1beta1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-12T16:51:26-03:00 with ADDON_CHECK_INTERVAL_SEC=30 ==
INFO: == Entering periodical apply loop at 2021-02-12T16:51:26-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T16:51:27-03:00 ==
INFO: == Reconciling with deprecated label ==
INFO: == Reconciling with addon-manager label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/dashboard created
INFO: == Kubernetes addon reconcile completed at 2021-02-12T16:51:28-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T16:51:56-03:00 ==
INFO: == Reconciling with deprecated label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Reconciling with addon-manager label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/dashboard unchanged
INFO: == Kubernetes addon reconcile completed at 2021-02-12T16:51:59-03:00 ==

------------- New manager + new kubectl from here, old manifests ----------------

./kube-addons-main.sh 
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist networking.k8s.io/v1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-12T16:57:23-03:00 with ADDON_CHECK_INTERVAL_SEC=30 ==
Client Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-alpha.3.164+528390348ddcf8-dirty", GitCommit:"528390348ddcf873ca47df9f12900685dac585a4", GitTreeState:"dirty", BuildDate:"2021-02-12T19:56:01Z", GoVersion:"go1.15.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-21T01:11:42Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
INFO: == Entering periodical apply loop at 2021-02-12T16:57:24-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T16:57:24-03:00 ==
INFO: == Reconciling with deprecated label ==
INFO: == Reconciling with addon-manager label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/dashboard unchanged
INFO: == Kubernetes addon reconcile completed at 2021-02-12T16:57:27-03:00 ==
```

**Comment by rikatz** [comment]
* Scenario 3: new addon manager + extensions/v1beta1 manifests → new addon manager + networking.k8s.io/v1 manifests
(from the previous tests, just updated the manifests!)

Once there's a new manifest, it gets re-applied from the first pass and then unchanged on later pass
 ```
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist networking.k8s.io/v1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-12T17:03:09-03:00 with ADDON_CHECK_INTERVAL_SEC=30 ==
Client Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-alpha.3.164+528390348ddcf8-dirty", GitCommit:"528390348ddcf873ca47df9f12900685dac585a4", GitTreeState:"dirty", BuildDate:"2021-02-12T19:56:01Z", GoVersion:"go1.15.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-21T01:11:42Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
INFO: == Entering periodical apply loop at 2021-02-12T17:03:10-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T17:03:10-03:00 ==
INFO: == Reconciling with deprecated label ==
INFO: == Reconciling with addon-manager label ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
ingress.extensions/dashboard unchanged
INFO: == Kubernetes addon reconcile completed at 2021-02-12T17:03:13-03:00 ==

------------ New Manifest from here ------------------

INFO: == Kubernetes addon ensure completed at 2021-02-12T17:03:40-03:00 ==
INFO: == Reconciling with addon-manager label ==
<----- Apparently Ingress gets re-applied from here ---->
INFO: == Kubernetes addon reconcile completed at 2021-02-12T17:03:43-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T17:04:10-03:00 ==
INFO: == Reconciling with deprecated label ==
INFO: == Reconciling with addon-manager label ==
ingress.networking.k8s.io/dashboard unchanged
INFO: == Kubernetes addon reconcile completed at 2021-02-12T17:04:13-03:00 ==
```

**Comment by rikatz** [comment]
* Scenario 4: old addon manager + networking.k8s.io/v1 manifests → new addon manager + networking.k8s.io/v1 manifests
After addonmanager + kubectl are updated, nothing is changed

```
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist extensions/v1beta1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-12T17:21:31-03:00 with ADDON_CHECK_INTERVAL_SEC=30 ==
Client Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-alpha.3.164+528390348ddcf8-dirty", GitCommit:"528390348ddcf873ca47df9f12900685dac585a4", GitTreeState:"dirty", BuildDate:"2021-02-12T20:13:31Z", GoVersion:"go1.15.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-21T01:11:42Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
INFO: == Entering periodical apply loop at 2021-02-12T17:21:31-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T17:21:32-03:00 ==
INFO: == Reconciling with deprecated label ==
INFO: == Reconciling with addon-manager label ==
ingress.networking.k8s.io/dashboard created
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Kubernetes addon reconcile completed at 2021-02-12T17:21:33-03:00 ==
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Reconciling with addon-manager label ==
ingress.networking.k8s.io/dashboard unchanged
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
INFO: == Kubernetes addon reconcile completed at 2021-02-12T17:22:04-03:00 ==

--------------- Update addonmanager ---------------

./kube-addons-main.sh 
INFO: == Generated kubectl prune whitelist flags: --prune-whitelist core/v1/ConfigMap --prune-whitelist core/v1/Endpoints --prune-whitelist core/v1/Namespace --prune-whitelist core/v1/PersistentVolumeClaim --prune-whitelist core/v1/PersistentVolume --prune-whitelist core/v1/Pod --prune-whitelist core/v1/ReplicationController --prune-whitelist core/v1/Secret --prune-whitelist core/v1/Service --prune-whitelist batch/v1/Job --prune-whitelist batch/v1beta1/CronJob --prune-whitelist apps/v1/DaemonSet --prune-whitelist apps/v1/Deployment --prune-whitelist apps/v1/ReplicaSet --prune-whitelist apps/v1/StatefulSet --prune-whitelist networking.k8s.io/v1/Ingress  ==
INFO: == Kubernetes addon manager started at 2021-02-12T17:26:25-03:00 with ADDON_CHECK_INTERVAL_SEC=30 ==
Client Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-alpha.3.164+528390348ddcf8-dirty", GitCommit:"528390348ddcf873ca47df9f12900685dac585a4", GitTreeState:"dirty", BuildDate:"2021-02-12T20:26:20Z", GoVersion:"go1.15.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-21T01:11:42Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
INFO: == Entering periodical apply loop at 2021-02-12T17:26:26-03:00 ==
INFO: == Kubernetes addon ensure completed at 2021-02-12T17:26:26-03:00 ==
INFO: == Reconciling with deprecated label ==
INFO: == Reconciling with addon-manager label ==
ingress.networking.k8s.io/dashboard unchanged
INFO: == Kubernetes addon reconcile completed at 2021-02-12T17:26:29-03:00 ==
```

**Comment by rikatz** [comment]
Scenarios:
1. old addon manager + extensions/v1beta1 manifests → old addon manager + networking.k8s.io/v1 manifests
2. old addon manager + extensions/v1beta1 manifests → new addon manager + extensions/v1beta1 manifests
3. new addon manager + extensions/v1beta1 manifests → new addon manager + networking.k8s.io/v1 manifests
4. old addon manager + networking.k8s.io/v1 manifests → new addon manager + networking.k8s.io/v1 manifests


| Scenario  | Result | Logs |
| --------- | ------------- | ------------- | 
|1 | Object gets re-applied on first pass and then unchanged  | [here](https://github.com/kubernetes/kubernetes/pull/98806#issuecomment-778416780) |
|2  | Nothing gets prunned/reapplied | [here](https://github.com/kubernetes/kubernetes/pull/98806#issuecomment-778422571) |
|3  | Object gets re-applied on first pass and then unchanged  | [here](https://github.com/kubernetes/kubernetes/pull/98806#issuecomment-778425157)|
|4  | Nothing gets prunned/reapplied | [here](https://github.com/kubernetes/kubernetes/pull/98806#issuecomment-778436629)|


/assign @liggitt

**Comment by liggitt** [comment]
Those results look great, thanks for running through them.

**Comment by liggitt** [comment]
this change looks good to me. cc @spencer-p for guidance on incrementing the addon manager version and getting the built manifest updated so this change is effective. There are several changes in progress that might need traffic direction:

* https://github.com/kubernetes/kubernetes/pull/98974
* https://github.com/kubernetes/kubernetes/pull/98968

**Comment by rikatz** [comment]
@MrHohn when you have some time, release managers/issue triagers pinged us on the original @liggitt issue about the issue that this PR fixes.

Do we plan to have this merged for v1.21? Otherwise we need to remove the milestone from issue and PR :)

Thanks!

**Comment by liggitt** [comment]
it needs to be merged for 1.21 to be forward compatible with 1.22 API servers

**Comment by MrHohn** [comment]
Sorry for the delay, thanks for taking the time doing all the recommended tests!

The changes look good to me, the next steps in my mind:
1. Get this PR and #98968 merged.
2. Push new images to the staging repo (I can help with this).
3. Promote the staging images to prod (similar to https://github.com/kubernetes/k8s.io/pull/1309).
4. Update the addon-manager manifests (similar to https://github.com/kubernetes/kubernetes/pull/98974).

**Review by MrHohn — APPROVED** [review]
/approve

**Comment by MrHohn** [comment]
Okay, since this also makes change to kubectl, we might need to wait for a new kubectl version and bump it here:
https://github.com/kubernetes/kubernetes/blob/ad1e2ffab00960bdc60b8dd61b2e4445665bf918/cluster/addons/addon-manager/Makefile#L19

**Comment by rikatz** [comment]
yes, but this might be the chicken/egg thing, right? because we need to merge the kubectl part here, and then change the Makefile :)

Let me get some sig/cli approval here

**Comment by liggitt** [comment]
the addon manager change sets the default prune list, so it doesn't matter what kubectl defaults to. I don't think we have to wait for a new kubectl

**Comment by MrHohn** [comment]
Oh right, @liggitt is correct that doesn't really matter for addon manager.

**Review by soltysh — APPROVED** [review]
/approve
sig-cli bits

**Comment by rikatz** [comment]
@liggitt @MrHohn I've spoken with @soltysh and the PR is approved, he just said that the LGTM and holding or now it's up to you :)

Let me know if you need something else from me on this PR, otherwise I'm moving forward with the other stuffs :)

**Comment by MrHohn** [comment]
Thanks, let's get this merged and keep other bits rolling.
/lgtm

**Pull request merged**
